src/NHibernate/Intercept/AbstractFieldInterceptor.cs

```diff
@@ -128,39 +128,43 @@ public object Intercept(object target, string fieldName, object value, bool sett
 				return InvokeImplementation;
 			}
 
-			if (IsInitializedField(fieldName))
+			if (IsUninitializedProperty(fieldName))
 			{
-				return value;
-			}
+				if (session == null)
+				{
+					throw new LazyInitializationException(EntityName, null, string.Format("entity with lazy properties is not associated with a session. entity-name:'{0}' property:'{1}'", EntityName, fieldName));
+				}
+				if (!session.IsOpen || !session.IsConnected)
+				{
+					throw new LazyInitializationException(EntityName, null, string.Format("session is not connected. entity-name:'{0}' property:'{1}'", EntityName, fieldName));
+				}
 
-			if (session == null)
-			{
-				throw new LazyInitializationException(EntityName, null, string.Format("entity with lazy properties is not associated with a session. entity-name:'{0}' property:'{1}'", EntityName, fieldName));
-			}
-			if (!session.IsOpen || !session.IsConnected)
-			{
-				throw new LazyInitializationException(EntityName, null, string.Format("session is not connected. entity-name:'{0}' property:'{1}'", EntityName, fieldName));
+				return InitializeField(fieldName, target);
 			}
 
-			if (IsUninitializedProperty(fieldName))
+			if (!IsUninitializedAssociation(fieldName))
 			{
-				return InitializeField(fieldName, target);
+				return value;
 			}
 
-			if (value.IsProxy() && IsUninitializedAssociation(fieldName))
+			if (value is INHibernateProxy nhproxy)
 			{
-				var nhproxy = value as INHibernateProxy;
 				value = InitializeOrGetAssociation(nhproxy, fieldName);
-				// Set the property value in order to be accessible when the session is closed
-				var persister = session.Factory.GetEntityPersister(entityName);
-				persister.SetPropertyValue(
-					target,
-					persister.EntityMetamodel.BytecodeEnhancementMetadata.UnwrapProxyPropertiesMetadata.GetUnwrapProxyPropertyIndex(fieldName),
-					value);
+				if (session?.Factory.IsClosed == false)
+				{
```

Comment:
Is this a fix for another bug?

Reply:
No. Open session was required before fix. I moved this check specifically to lazy property initialization case (see https://github.com/nhibernate/nhibernate-core/pull/3347/files#diff-ee99d915582e6c18e9612d53afa6360dedd6ce7fc5b025de855cf09a79a3a4cdR131-R140)

Now when proxy is initialized we can have null or closed session/factory.